text/3199-unchecked-lifetime.md

```diff
@@ -0,0 +1,171 @@
+```rust
+struct ArrayIter<T> {
+    buffer: [T; 32]
+    iter: std::slice::Iter<'static, T>
+}
+```
+What if T is not `'static`? using the static lifetime here restricts our generic parameter T to being 'static, which is a concession we may not be ok with making.
+
+So how do we get all of the above? We use the "unchecked lifetime" `'?`
+```rust
+struct ArrayIter<T> {
+    buffer: [T; 32]
+    iter: std::slice::Iter<'?, T>
+}
+```
+
+Note that, like `'static`, `'?` is allowed to appear in struct definitions without being declared. This is because the unchecked lifetime instructs the borrow checker to treat any references with this lifetime like raw pointers. This is very unsafe of course, so as a tradeoff, dereferencing a reference with the unchecked lifetime is unsafe, and `'?` is not acceptable when `'a` is expected.
+
+In general using replacing a real lifetime with `'?` should be thought of as a similar transformation to replacing a reference with a pointer. If you are doing it, you are doing it because safe rust does not allow for the type of code you are trying to write, and you're trying to encapsulate the unsafe into a compact part of your code.
+
+If you try to call a method whose arguments or return value include `'?`, that call will need to be wrapped in unsafe, because you are asserting that you know those references are valid despite the borrow checker not knowing.
+
+The addition of the `'?` lifetime also means the addition of two new reference types, `&'? T` and `&'? mut T`. These are in a sense halfway in between references and pointers. Dereferencing them is unsafe. Static references can be coerced into normal references, which can be coerced into unchecked-lifetime references, which can be coerced into raw pointers. The crucial difference between `&'? T` and `*const T` is that it is considered unsound for `&'? T` to be unaligned at any time, instead of only at the time of dereference for raw pointers.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The unchecked lifetime is unique in a few ways:
+- It is outlived by all other lifetimes (in the same way that `'static` outlives all other lifetimes).
+- The check that a reference does not outlive borrowed content is skipped when the borrowed content has `'?` lifetime. (This is the check that goes *unchecked*).
+- It cannot be used where a normal reference is expected.
+- assigning to or reading from `&'? T` and `&'? mut T` is unsafe
+
+Let's examine some of the implications of these features.
+
+It is safe in general to store into a value expecting `'?`, just like with raw pointers
+
+```rust
+struct A {
+    val: &'? usize
+}
+
+impl A {
+    fn store(&self, some_ref: &usize) {
+        self.val = some_ref; // assigning to val, not through val.
+    }
+}
+```
+An important but subtle point here is that it is only legal to have `&self` because of rule 2. Without it all references to A would outlive their borrowed content.
```

Comment:
```suggestion
An important but subtle point here is that it is only legal to have `&mut self` because of rule 2. Without it all references to A would outlive their borrowed content.
```